src/template/oidrequest.tt

```diff
@@ -0,0 +1,483 @@
+    if (NDIS_STATUS_PENDING != NdisStatus)
+    {
+        NdisCopyOidResult(OidRequest, Clone);
+        NdisFreeCloneOidRequest(NdisFilterHandle, Clone);
+    }
+
+    return NdisStatus;
+}
+
+typedef enum NDIS_OID_REQUEST_COMPLETION_KIND_t
+{
+    NdisOidRequestCompletionKindPassthrough,
+    NdisOidRequestCompletionKindLocal,
+} NDIS_OID_REQUEST_COMPLETION_KIND;
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+inline
+NDIS_OID_REQUEST_COMPLETION_KIND
+NdisFPassthroughOidRequestComplete(
+    _In_ NDIS_HANDLE NdisFilterHandle,
+    _In_ NDIS_OID_REQUEST *OidRequest,
+    _In_ NDIS_STATUS CompletionStatus)
+/*++
+
+Routine Description:
+
+    Handles the completion of an OID request
+
+    Example usage:
+
+        void MyFilterOidRequestComplete(
+            NDIS_HANDLE filterModuleContext,
+            NDIS_OID_REQUEST *oid,
+            NDIS_STATUS completionStatus)
+        {
+            MY_FILTER *filter = (MY_FILTER*)filterModuleContext;
+            NdisFPassthroughOidRequestComplete(
+                filter->ndisHandle, oid, completionStatus);
+        }
+
+    This completion handler may only be used with OIDs that were driven from
+    NdisFPassthroughOidRequest, NdisFOidRequestAsync, or
+    NdisFOidRequestAndWait. If your LWF issues OIDs in other ways, you must
+    ensure these OIDs are not passed to NdisFPassthroughOidRequestComplete.
+
+Arguments:
+
+    NdisFilterHandle - The NDIS handle of your filter module
+
+    OidRequest - The OID request that is being completed
+
+    CompletionStatus - The resulting status of the OID request
+
+Return Value:
+
+    In some cases, you might need to know where the OID request came from. This
+    information is not required in all cases, and it is often correct to ignore
+    the return value of this routine.
+
+    NdisOidRequestCompletionKindPassthrough - the OID request was issued by
+         NdisFPassthroughOidRequest
+
+    NdisOidRequestCompletionKindLocal - the OID request was issued by
+         NdisFOidRequestAsync or NdisFOidRequestAndWait
+
+--*/
+{
+    NDIS_OID_REQUEST_SOURCE_RESERVED *Context =
+        (NDIS_OID_REQUEST_SOURCE_RESERVED*)&OidRequest->SourceReserved[0];
+
+    if (NULL == Context->OriginalRequest)
+    {
+        WriteRelease(&Context->Completion.Status, CompletionStatus);
+        if (NULL != Context->Completion.WaitEvent)
+        {
+            KeSetEvent(Context->Completion.WaitEvent, IO_NO_INCREMENT, FALSE);
+        }
+
+        return NdisOidRequestCompletionKindLocal;
+    }
+    else
+    {
+        NDIS_OID_REQUEST *OriginalRequest = Context->OriginalRequest;
+        NdisCopyOidResult(OriginalRequest, OidRequest);
+        NdisFreeCloneOidRequest(NdisFilterHandle, OidRequest);
+        NdisFOidRequestComplete(NdisFilterHandle, OriginalRequest, CompletionStatus);
+        return NdisOidRequestCompletionKindPassthrough;
+    }
+}
+
+_IRQL_requires_(PASSIVE_LEVEL)
+inline
+NDIS_STATUS
+NdisFOidRequestAsync(
```

Comment:
Maybe NdisFIssueOidRequestAsync? The passthrough functions allocate a clone and these ones don't. I understand the difference, but it may not be so clear for someone new.